streamz/sources.py

```diff
@@ -131,35 +129,148 @@ class from_kafka(Source):
+            val = self.do_poll()
+            if val:
+                yield self._emit(val)
+            else:
+                yield gen.sleep(self.poll_interval)
+
+    def start(self):
+        import confluent_kafka as ck
+        import distributed
+        finalize = distributed.compatibility.finalize
+        if self.loop is not None:
+            # need some IOLoop to run on
+            self.loop.add_callback(self.poll_kafka)
+        self.consumer = ck.Consumer(self.cpars)
+        self.consumer.subscribe(self.topics)
+
+        def close(ref):
+            ob = ref()
+            if ob is not None and ob.consumer is not None:
+                ob.consumer.close()
+
+        finalize(self, close, weakref.ref(self))
+
+    def _close_consumer(self):
+        if self.consumer is not None:
+            self.consumer.close()
+            self.consumer = None
+
+
+@Stream.register_api(staticmethod)
+class from_kafka_batched(Source):
+    def __init__(self, topic, consumer_params, poll_interval=1., npartitions=1,
+                 start=True, min_batch=100, dask_client=None, **kwargs):
+        self.cpars = consumer_params
+        self.consumer = None
+        self.topic = topic
+        self.npartitions = npartitions
+        self.min_batch = min_batch
+        self.positions = [0] * npartitions
+        self.poll_interval = poll_interval
+        self.client = dask_client
+        super(from_kafka_batched, self).__init__(ensure_io_loop=True, **kwargs)
+        if start:
+            self.start()
+
+    def do_poll(self):
+        """Check each partition for messages
+
+        Returns either a list futures of tasks to download, of self.client is
+        defined, or does the downloads immediately, returning lists of values
+        for each partition.
+
+        Will only consider partitions where the current max offset is at least
+        min_batch since the last time messages were fetched.
+        """
+        import confluent_kafka as ck
+        consumer = ck.Consumer(self.cpars)
+        out = []
+        for partition in range(self.npartitions):
+            tp = ck.TopicPartition(self.topic, partition, 0)
+            low, high = consumer.get_watermark_offsets(tp, timeout=0.1)
+            pos = self.positions[partition]
+            if self.min_batch is None or high - pos > self.min_batch:
+                self.positions[partition] = high
+                if self.client is not None:
+                    future = self.client.submit(
+                        get_message_batch, self.cpars, self.topic, partition,
+                        pos, high - 1)
+                    out.append(future)
+                else:
+                    out.append(get_message_batch(
+                        self.cpars, self.topic, partition, pos, high - 1))
+        return out
+
+    @gen.coroutine
+    def poll_kafka(self):
+        while True:
+            out = self.do_poll()
+            if any(out):
+                yield self._emit(out)
```

Comment:
I was wondering about this. I don't know which is preferable, it's a general library design decision on whether we want atomicity or to reduce the number of emit calls. In common cases I don't suppose there are many partitions, but we should still check for None, where there is no new data.

Reply:
Semantically I think of these data sources as being streams of records, not streams of batches of records.  When considering efficiency on high volume topics I think we would end up switching to the dask-friendly batched solution, even if dask is not used.  I agree that we shouldn't emit when no data arrives, I think that this will happen naturally if `self.do_poll()` returns an empty list?